### ISSUE TITLE
Investigate SQLAlchemy's nested transations

### ISSUE THREAD

**Issue by jameshiebert** [pacificclimate/crmprtd#9]
There is some evidence that suggests our nested transactions are not behaving properly. The ``moti_infill.py`` script is reporting that it's making insertions, however those insertions never actually appear in the database.

The transaction management in this package is "complicated". We haven't pinned our version of ``SQLAlchemy`` in the ``requirements.txt`` so there is a real possibility that the behaviour of nested transactions has changed out from under us. We need to:

1. Upgrade and pin to the latest stable version of ``SQLAlchemy`` and
1. Ensure that the nested transactions are behaving the way that they should

**Comment by rod-glover** [comment]
I think you must mean `moti_infill_insert.py`.

**Comment by rod-glover** [comment]
Short version:

- All database activities for all contents (observations) in all files processed are rooted in one session.
- That session (and not a nested session!) is rolled back if any two of the observations are duplicates (uniqueness enforced on history, variable, time)
- Therefore if there is any duplicate, none of the observations in the entire run are added to the database.

Fortunately, this won't be hard to fix.

**Comment by rod-glover** [comment]
Addendum: I don't think this is necessarily related to a change in how SQLAlchemy handles nested transactions. I'd guess the difference is that we are encountering duplicate observations in our input files.
I'm pretty sure duplication can arise from an observation being already in the database from a previous run, as well as a duplication in the files being processed in the present run.

**Comment by jameshiebert** [comment]
Ahhhh, OK it's good that you have taken a closer look at this. Since we have written scripts for different networks at different times, and because each network has a slightly different data format, our error handing and transaction management strategies have varied widely over the years/networks.

I think where we (I) have gone wrong, is assuming that the `moti_infill_insert.py` script actually handles duplicate observations on an individual basis, since that's what some of our other insertion scripts do.

Would you be able to migrate the handling of duplicates from other scripts into the `moti_infill_insert.py` script? We currently have two strategies for this.

1) wrap each insert into a nested transaction and rollback on error
1) insert everything and on error, rollback and do a binary split and retry (we do this in ENV-AQN)

There's currently no evidence that 2) is faster, but I believe in my heart that it should be :)

**Comment by rod-glover** [comment]
Yeah. As it turns out, my problem analysis above isn't exactly right, but it's close. The nested transaction isn't quite being used correctly, and it's consequently rolling back too much. I've got a fix, but I don't like the smell of it. SQLAlchemy documentation here isn't all one could wish it to be. Should be out of the woods shortly.

Strategy 1 is what `moti_infill_insert.py` originally used and I am still using. I think it can be made good. 
Strategy 2 might be faster, but it sounds like more work and more complex code. Premature optiimization?

**Comment by rod-glover** [comment]
There appears to be a contradiction between what the documentation says and what we can write that works. (Caveat: Our code is not a minimal example, so there could be something non-obvious going on in it that confuses matters.)

[Documentation](http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#session-begin-nested) says:

> begin_nested() may be called any number of times, which will issue a new SAVEPOINT with a unique identifier for each call. For each begin_nested() call, a corresponding rollback() or commit() must be issued. (But note that if the return value is used as a context manager, i.e. in a with-statement, then this rollback/commit is issued by the context manager upon exiting the context, and so should not be added explicitly.)

1. Our original failing code:
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```
That code goes against the documentation by adding an explicit rollback (which is supposed to be provided by the context manager returned by `begin_nested`). And it fails because it provably (testably) rolls back all the objects inserted up to that point, but not those after.

2. Documentation-compliant failing code (remove supposedly spurious explicit rollback):
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                skips += 1
```
That code fails with the following error message:
```
E                       sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (psycopg2.IntegrityError) duplicate key value violates unique constraint "time_place_variable_unique"
E                       DETAIL:  Key (obs_time, history_id, vars_id)=(2012-01-02 01:00:00, 1, 1) already exists.
E                        [SQL: 'INSERT INTO crmp.obs_raw (obs_time, mod_time, datum, vars_id, history_id) VALUES (%(obs_time)s, %(mod_time)s, %(datum)s, %(vars_id)s, %(history_id)s) RETURNING crmp.obs_raw.obs_raw_id'] [parameters: {'obs_time': datetime.datetime(2012, 1, 2, 1, 0), 'history_id': 1, 'vars_id': 1, 'datum': -2.417, 'mod_time': datetime.datetime(2017, 11, 30, 0, 9, 58, 960640)}]
```

3. Failing alternative code that uses explicit commit and rollback, without a context manager:
```
            try:
                sesh.begin_nested()
                sesh.add(o)
                sesh.commit()
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```
That code also fails because it rolls back everything inserted before the rollback, but not the objects inserted after.

4. Working code that uses both context manager and explicit commit and rollback, contrary to recommendation in documenation:
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                sesh.commit()
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```

Questions: WTF? What am I missing here?

**Comment by jameshiebert** [comment]
What the heck?! Yes, number 2 definitely looks like the *correct* answer, but if that's not working, I'd raise an issues against SQLAlchmey. (You *are* using the same version as the documentation, yes?)

**Comment by rod-glover** [comment]
I am using 1.1.15, whose documentation for this is identical. SQLAlchemy just rolled out 1.2.0b3, which is labelled PRE-RELEASE. It's what you get for *latest*, but again it's supposedly the same for this feature. I'll try installing 1.2 and see what happens, but I don't expect any difference.

**Comment by rod-glover** [comment]
You'll note that the error message for 2 advises adding the rollback call, but that actually causes the error. This is most frustrating and surprising considering that they offer this example, which is very similar to our case (but much simpler), of how to use nested transactions to handle integrity failures for individual operations:
```
for record in records:
    try:
        with session.begin_nested():
            session.merge(record)
    except:
        print("Skipped record %s" % record)
session.commit()```

**Comment by rod-glover** [comment]
Wait ... the latest release available via PyPI is 1.15, and its release date (Nov 3) is later than 1.2.0b3 (Oct 13). Their documentation website is a bit goofy, or at the least misleading. I think it is documenting 1.1.15, and I think that 1.1.15 is the latest stable release.
